<h1>📚 Maestría en Inteligencia Artificial Aplicada – 3er Semestre</h1>

<h3>Asignatura: Procesamiento de Lenguaje Natural</h3>

<h4>Práctica Nro. 5 - Generación de Texto con Modelos GPT</h4>


<hr style="width:60%;">

<h2>👨‍🎓 Estudiantes</h2>
<ul style="list-style:none; padding:0; font-size:18px;">
    <li>Claudia Martínez</li>
    <li>Sebastián Murillas</li>
    <li>Mario J. Castellanos</li>
    <li>Enrique Manzano</li>
    <li>Octavio Guerra</li>
</ul>

<hr style="width:60%;">

<h3>📅 Fecha: Septiembre 21, 2025</h3>

</center>

# Generación de Texto con modelos GPT

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/semurillas/NLP_MIAA_252/blob/main/Sesion%205/Practice/1-text-generation.ipynb)

En este Notebook harémos uso de un modelo tipo GPT-2 pre-entrenado en idioma inglés que utilizaremos para generar texto a partir de un contexto inicial que proveerémos. Luego, harémos fine tuning a este modelo con un dataset de chistes (jokes) en Inglés y observar como cambia la generación de texto en función del dataset que utilicemos.

#### Referencias
- Dataset: https://www.kaggle.com/datasets/abhinavmoudgil95/short-jokes
- [Improving Language Understanding by Generative Pre-Training](https://cdn.openai.com/research-covers/language-unsupervised/language_understanding_paper.pdf)
- [Natural Language Processing with Transformers: Building Language Applications With Hugging Face](https://www.amazon.com/Natural-Language-Processing-Transformers-Applications/dp/1098103246)
- [GPT2 Spanish](https://huggingface.co/DeepESP/gpt2-spanish)
- [Fine-Tuning GPT-2 with Hugging Face Transformers: A Complete Guide](https://www.cohorte.co/blog/fine-tuning-gpt-2-with-hugging-face-transformers-a-complete-guide)

## **1. Chequeo de ambiente y Manejo de "Warnings"**

Iniciamos verificando si estamos trabajando en "Google Colab" y deshabilitar algun mensaje de "warning" durante la ejecución de las celdas de código Python.

In [ ]:
import importlib.metadata    # Importación del modulo de metadata de la libreria importlib
import warnings               # Importación de la libreria warnings

# Uso de la libreria importlib.metadata para obtener los paquetes
# instalados en el Notebook
installed_packages = [dist.metadata['Name'].lower() for dist in importlib.metadata.Distribution.discover()]

# Chequeando que Google Colab este entre los paquetes instalados
# para confirmar si es el ambiente de trabajo usado.
IN_COLAB = 'google-colab' in installed_packages

# Ignorar algun Warning generado durante la ejecución del Notebook
warnings.filterwarnings('ignore')

## **2. Preparación del Entorno de Google Colab**


El archivo ejecutado a continuación en el entorno del Notebook:  `requirements.txt` incluye las siguientes librerías:

- **Core tools:** manejo de instalación y empaquetado (`pip`, `setuptools`, `wheel`).
- **NLP:** procesamiento de lenguaje natural con *spaCy*, *NLTK*, *Transformers*, *Datasets* y *Sentence-Transformers*.
- **ML / DL:** frameworks de aprendizaje profundo (*PyTorch*, *Lightning*), soporte de audio/imagen y utilidades de entrenamiento (*TensorBoard*, *Accelerate*, *Evaluate*).
- **Data / Visualización:** análisis y gráficos con *Pandas*, *Matplotlib*, *Seaborn*, *Statsmodels*, *TQDM* y *Bokeh*.
- **Utils:** herramientas para interacción y despliegue (*Gradio*, *Ollama*, *Websockets*, *HTTPX*, *python-docx*).
- **Compatibilidad:** versiones fijas de *NumPy*, *scikit-learn*, *filelock*, *fsspec* y *gcsfs* para evitar conflictos.
- **Extras (Colab):** soporte extendido de *OpenCV* y *UMAP-learn* para reducción de dimensionalidad y visión por computador.
- **Kaggle API:** descarga directa de datasets desde Kaggle.

Este conjunto de librerías nos asegura un entorno reproducible y estable para entrenar, evaluar y desplegar modelos de PLN y en este caso el requerido para generación de Texto con Modelos de ChatGPT


In [ ]:
!test '{IN_COLAB}' = 'True' &&  pip install -r "https://raw.githubusercontent.com/semurillas/NLP_MIAA_252/refs/heads/main/Sesion%205/Practice/requirements.txt?nocache=$(date +%s)"

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 3.6 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of gcsfs to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 98.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.3/527.3 kB 41.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 779.1/779.1 MB 1.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.0/7.0 MB 124.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 93.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 324.4/324.4 kB 30.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 62.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 111.3 MB/s

## **3. Importación de librerías principales**

A continuación realizamos la importación de las librerías necesarias para trabajar con modelos de Procesamiento de Lenguaje Natural.  
Se incluyen **PyTorch** y sus módulos para la construcción de redes neuronales, y la librería **Transformers de Hugging Face** para cargar modelos preentrenados, tokenizadores y herramientas de entrenamiento como `Trainer` y `TrainingArguments`.  
También integramos **Datasets** para acceder a conjuntos de datos, y librerías clásicas de análisis como **NumPy** y **Pandas** para la manipulación de información. Finalmente, añadimos utilidades como `os` y `typing` que facilitan la gestión del entorno y aportan claridad en el código.


In [ ]:
# Utilidades del sistema
import os

# Deep Learning
import torch
import torch.nn as nn

# NLP - Hugging Face Transformers
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    DataCollatorForLanguageModeling,
    Trainer,
    TrainingArguments
)
from transformers.tokenization_utils_base import PreTrainedTokenizerBase

# uso de OpenAI para importar Modelo GPT-4
from openai import OpenAI

# Datasets
from datasets import load_dataset

# Manipulación de Datos
import numpy as np
import pandas as pd

# Tipado (para claridad en funciones y clases)
from typing import Optional, Tuple

## **4. Generative pre-training Transformer - GPT**

![](../assets/gpt.png)

Los modelos tipo GPT, introducidos por Radfor, et.al., de OpenAI, al igual que los modelos BERT, hacen uso extensivo de la arquitectura de transformers como hemos estado viendo. Las diferencias claves se podrían resumir en:

1. GPT utiliza bloques de **Transformer Decoder** encadenados, mientras que el modelo BERT utiliza bloques de *Transformer Encoder*
2. GPT se centra en la generación de texto basado en un contexto, la tarea principal es la predicción del siguiente token en la secuencia, mientras que BERT se centra en el completado de partes de una secuencia, en función de un contexto anterior y posterior a la secuencia de entrada. Entonces BERT se centra en la construicción de representación de lenguage, mientras que GPT se centra en la generación de texto en función de un contexto.

Sin embargo, ambos se basan en la misma premisa de pre-entrenar el modelo en tareas no-supervisadas o semi-supervisadas para que el modelo aprenda las representaciones semánticas del lenguage y luego al modelo se le pueda hacer fine tuning a tareas posteriores.

## **5. Selección de GPU o CPU**

 En la siguiente celada de código, seleccionamos de manera automática GPU (cuda) si está disponible, de lo contrario se usa CPU. Es recomendable el uso de GPU, para el entrenamiento de Modelos, como es el propósito mas adelante en el codigo desarrollado en este Notebook.

In [ ]:
device = "cuda" if torch.cuda.is_available() else "cpu"

## **6. Carga y Preparación de los Modelos**

Vamos a trabajar con dos (2) Modelos:

- **GPT-2**
y
- **GPT-Neo**

La idea es comparar que tan buenos son en la generación de textos sobre chistes (jokes) en Inglés, usando para entrenamiento y pruebas un dataset que ya describiremos mas adelante con un listado significativo de chistes en este idioma.

### **6.1. Carga y Preparación del Modelo GPT-2**

En la siguiente celda configuramos el modelo de lenguaje **GPT-2** junto con su tokenizador.  
Primero cargamos los recursos preentrenados, luego definimos un token de padding para evitar errores en el entrenamiento, y ajustamos los embeddings en caso de cambios en el vocabulario. Finalmente, movemos el modelo al dispositivo disponible (**GPU o CPU**) para su posterior uso en generación de texto.

#### Referencias
- [GPT-2](https://huggingface.co/docs/transformers/en/model_doc/gpt2)

In [ ]:
# Definicón del uso del Modelo GPT-2 preentrenado
model_gpt2 = "gpt2"

# Cargamos el tokenizador asociado al modelo
tokenizer_gpt2 = AutoTokenizer.from_pretrained(model_gpt2)

# Cargamos el modelo de lenguaje causal preentrenado (GPT-2)
model_gpt2 = AutoModelForCausalLM.from_pretrained(model_gpt2)

# Definimos un token de padding usando el mismo que EOS (fin de secuencia)
tokenizer_gpt2.pad_token = tokenizer_gpt2.eos_token
model_gpt2.config.pad_token_id = tokenizer_gpt2.pad_token_id

# Ajustamos las dimensiones de los embeddings en caso de que cambie el tokenizador
model_gpt2.resize_token_embeddings(len(tokenizer_gpt2))

# Movemos el modelo al dispositivo de cómputo disponible (GPU o CPU)
model_gpt2 = model_gpt2.to(device)

# Mostramos la configuración general del modelo cargado
print("GPT-2 cargado correctamente")
print(model_gpt2)


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

GPT-2 cargado correctamente
GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2SdpaAttention(
          (c_attn): Conv1D()
          (c_proj): Conv1D()
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D()
          (c_proj): Conv1D()
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)


In [ ]:
modules_gpt2 = [m for m, _ in model_gpt2.named_modules()]
modules_gpt2

['',
 'transformer',
 'transformer.wte',
 'transformer.wpe',
 'transformer.drop',
 'transformer.h',
 'transformer.h.0',
 'transformer.h.0.ln_1',
 'transformer.h.0.attn',
 'transformer.h.0.attn.c_attn',
 'transformer.h.0.attn.c_proj',
 'transformer.h.0.attn.attn_dropout',
 'transformer.h.0.attn.resid_dropout',
 'transformer.h.0.ln_2',
 'transformer.h.0.mlp',
 'transformer.h.0.mlp.c_fc',
 'transformer.h.0.mlp.c_proj',
 'transformer.h.0.mlp.act',
 'transformer.h.0.mlp.dropout',
 'transformer.h.1',
 'transformer.h.1.ln_1',
 'transformer.h.1.attn',
 'transformer.h.1.attn.c_attn',
 'transformer.h.1.attn.c_proj',
 'transformer.h.1.attn.attn_dropout',
 'transformer.h.1.attn.resid_dropout',
 'transformer.h.1.ln_2',
 'transformer.h.1.mlp',
 'transformer.h.1.mlp.c_fc',
 'transformer.h.1.mlp.c_proj',
 'transformer.h.1.mlp.act',
 'transformer.h.1.mlp.dropout',
 'transformer.h.2',
 'transformer.h.2.ln_1',
 'transformer.h.2.attn',
 'transformer.h.2.attn.c_attn',
 'transformer.h.2.attn.c_proj',
 'tran

## **Comentario sobre el Modelo GPT-2**
Este modelo se compone de un transformer causal con 12 capas (h0–h11) en esta versión base. Cada capa incluye:

- LayerNorm (normalización de capas) para estabilidad numérica.

- Self-Attention (attn) para capturar dependencias entre tokens.

- MLP (c_fc, c_proj) con activación NewGELUActivation para procesamiento no lineal.

El embedding de entrada (wte) y de posición (wpe) codifican tokens y posiciones.
La salida final del transformer se conecta a lm_head (una capa lineal) que proyecta al vocabulario de 50257 tokens.

### **6.2. Carga y Preparación del Modelo GPT-Neo**

En la siguiente celda configuramos el modelo de lenguaje **GPT-Neo** junto con su tokenizador.  
Primero cargamos los recursos preentrenados, luego definimos un token de padding para evitar errores en el entrenamiento, y ajustamos los embeddings en caso de cambios en el vocabulario. Finalmente, movemos el modelo al dispositivo disponible (**GPU o CPU**) para su posterior uso en generación de texto.

**GPT-Neo** es un modelo de lenguaje causal open-source desarrollado por EleutherAI, diseñado para generación de texto similar a GPT-3. Está disponible en varias versiones, desde 125M hasta 2.7B parámetros, y puede usarse para tareas de generación de texto, completado de prompts y experimentación con datasets personalizados. Es mucho mas moderno y puede generar mejores textos que GPT-2

#### Referencias
- [GPT-NEO](https://huggingface.co/EleutherAI/gpt-neo-1.3B)

In [ ]:
# -------------------- Definición del modelo GPT-Neo preentrenado --------------------
model_gptneo = "EleutherAI/gpt-neo-1.3B"

# Cargamos el tokenizador asociado al modelo
tokenizer_gptneo = AutoTokenizer.from_pretrained(model_gptneo)

# Cargamos el modelo de lenguaje causal preentrenado
model_gptneo = AutoModelForCausalLM.from_pretrained(model_gptneo)

# Definimos un token de padding usando el mismo que EOS (fin de secuencia)
tokenizer_gptneo.pad_token = tokenizer_gptneo.eos_token
model_gptneo.config.pad_token_id = tokenizer_gptneo.pad_token_id

# Ajustamos las dimensiones de los embeddings en caso de que cambie el tokenizador
model_gptneo.resize_token_embeddings(len(tokenizer_gptneo))

# Movemos el modelo al dispositivo de cómputo disponible (GPU o CPU)
model_gptneo = model_gptneo.to(device)

# Mostramos la configuración general del modelo cargado
print("GPT-Neo cargado correctamente")
print(model_gptneo)

GPT-Neo cargado correctamente
GPTNeoForCausalLM(
  (transformer): GPTNeoModel(
    (wte): Embedding(50257, 2048)
    (wpe): Embedding(2048, 2048)
    (drop): Dropout(p=0.0, inplace=False)
    (h): ModuleList(
      (0-23): 24 x GPTNeoBlock(
        (ln_1): LayerNorm((2048,), eps=1e-05, elementwise_affine=True)
        (attn): GPTNeoAttention(
          (attention): GPTNeoSelfAttention(
            (attn_dropout): Dropout(p=0.0, inplace=False)
            (resid_dropout): Dropout(p=0.0, inplace=False)
            (k_proj): Linear(in_features=2048, out_features=2048, bias=False)
            (v_proj): Linear(in_features=2048, out_features=2048, bias=False)
            (q_proj): Linear(in_features=2048, out_features=2048, bias=False)
            (out_proj): Linear(in_features=2048, out_features=2048, bias=True)
          )
        )
        (ln_2): LayerNorm((2048,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTNeoMLP(
          (c_fc): Linear(in_features=2048, out_features=8192, 

In [ ]:
modules_gptneo = [m for m, _ in model_gptneo.named_modules()]
modules_gptneo

['',
 'transformer',
 'transformer.wte',
 'transformer.wpe',
 'transformer.drop',
 'transformer.h',
 'transformer.h.0',
 'transformer.h.0.ln_1',
 'transformer.h.0.attn',
 'transformer.h.0.attn.attention',
 'transformer.h.0.attn.attention.attn_dropout',
 'transformer.h.0.attn.attention.resid_dropout',
 'transformer.h.0.attn.attention.k_proj',
 'transformer.h.0.attn.attention.v_proj',
 'transformer.h.0.attn.attention.q_proj',
 'transformer.h.0.attn.attention.out_proj',
 'transformer.h.0.ln_2',
 'transformer.h.0.mlp',
 'transformer.h.0.mlp.c_fc',
 'transformer.h.0.mlp.c_proj',
 'transformer.h.0.mlp.act',
 'transformer.h.0.mlp.dropout',
 'transformer.h.1',
 'transformer.h.1.ln_1',
 'transformer.h.1.attn',
 'transformer.h.1.attn.attention',
 'transformer.h.1.attn.attention.attn_dropout',
 'transformer.h.1.attn.attention.resid_dropout',
 'transformer.h.1.attn.attention.k_proj',
 'transformer.h.1.attn.attention.v_proj',
 'transformer.h.1.attn.attention.q_proj',
 'transformer.h.1.attn.attentio

## **Comentario sobre el Modelo GPT-Neo**

Este Modelo tiene 24 capas (h0–h23), el doble de la base de GPT-2, lo que permite modelar dependencias más complejas. Cada capa incluye:

- LayerNorm (ln_1 y ln_2) para estabilizar los cálculos.

- Self-Attention (attn.attention) con proyecciones de queries, keys y values (q_proj, k_proj, v_proj) y salida (out_proj).

- MLP (c_fc, c_proj) con activación NewGELUActivation para procesamiento no lineal.

Embeddings:

wte: embedding de tokens (50257 tokens × 2048 dimensiones).

wpe: embedding de posiciones (2048 × 2048 dimensiones).

La salida final (lm_head) proyecta a logits sobre el vocabulario.

## **7. Prueba de los Modelos GPT-2 y GPT-Neo con una generación sencilla de texto**

Veamos un ejemplo de generación simple de texto usando GPT-2 y GPT-Neo.

### **7.1. Usando GPT-2**

In [ ]:
# Definimos el texto inicial que vamos a usar como prompt para el modelo.
text = "Once upon time"

# Establecemos el número de tokens más probables que queremos mostrar.
best = 10

# Desactivamos el cálculo de gradientes porque solo queremos inferencia, no entrenamiento.
with torch.no_grad():

    # Tokenizamos el texto y lo convertimos en un tensor compatible con PyTorch,
    # luego lo movemos al dispositivo (CPU o GPU) que estemos usando.

    tokens = tokenizer_gpt2(text, return_tensors='pt')['input_ids'].to(device)

    # Imprimimos las dimensiones del tensor de entrada para entender su forma.
    print("Dimensiones de la entrada:", tokens.shape)

    # Pasamos los tokens por el modelo GPT-2 para obtener los logits de predicción.
    output = model_gpt2(input_ids=tokens)

    # Revisamos las dimensiones de la salida para verificar el tamaño del vocabulario
    # y la longitud de la secuencia procesada.
    print("Dimensiones de la salida:", output.logits.shape)

    # Tomamos solo los logits del último token de la secuencia,
    # ya que queremos predecir el siguiente token.
    output = output.logits[0, -1, :]

    # Mostramos la forma del vector que representa las predicciones del próximo token.
    print("Dimensiones del último token de la secuencia:", output.shape)

    # Aplicamos softmax para convertir los logits en probabilidades normalizadas.
    probs = torch.softmax(output, dim=-1)

    # Imprimimos las dimensiones del vector de probabilidades.
    print("Dimensiones de la probabilidad de los tokens:", probs.shape)

    # Ordenamos los tokens según su probabilidad de mayor a menor.
    sorted_probs = torch.argsort(probs, dim=-1, descending=True)

    # Mostramos los 'best' tokens más probables y su probabilidad en porcentaje.
    print({tokenizer_gpt2.decode(token): f"{prob.cpu().numpy() * 100:.2f}%"
           for token, prob in zip(sorted_probs[:best], probs[sorted_probs[:best]])})


Dimensiones de la entrada: torch.Size([1, 3])
Dimensiones de la salida: torch.Size([1, 3, 50257])
Dimensiones del último token de la secuencia: torch.Size([50257])
Dimensiones de la probabilidad de los tokens: torch.Size([50257])
{',': '46.95%', ' the': '10.24%', ' I': '3.43%', ' he': '3.41%', ' you': '2.93%', ' it': '2.86%', ' they': '2.32%', ' of': '2.19%', ' a': '1.72%', ' we': '1.37%'}


### **7.2. Usando GPT-Neo**

In [ ]:
# Definimos el texto inicial que vamos a usar como prompt para el modelo.
text = "Once upon time"

# Establecemos el número de tokens más probables que queremos mostrar.
best = 10

# Desactivamos el cálculo de gradientes porque solo queremos inferencia, no entrenamiento.
with torch.no_grad():

    # Tokenizamos el texto y lo convertimos en un tensor compatible con PyTorch,
    # luego lo movemos al dispositivo (CPU o GPU) que estemos usando.

    tokens = tokenizer_gptneo(text, return_tensors='pt')['input_ids'].to(device)

    # Imprimimos las dimensiones del tensor de entrada para entender su forma.
    print("Dimensiones de la entrada:", tokens.shape)

    # Pasamos los tokens por el modelo GPT-2 para obtener los logits de predicción.
    output = model_gptneo(input_ids=tokens)

    # Revisamos las dimensiones de la salida para verificar el tamaño del vocabulario
    # y la longitud de la secuencia procesada.
    print("Dimensiones de la salida:", output.logits.shape)

    # Tomamos solo los logits del último token de la secuencia,
    # ya que queremos predecir el siguiente token.
    output = output.logits[0, -1, :]

    # Mostramos la forma del vector que representa las predicciones del próximo token.
    print("Dimensiones del último token de la secuencia:", output.shape)

    # Aplicamos softmax para convertir los logits en probabilidades normalizadas.
    probs = torch.softmax(output, dim=-1)

    # Imprimimos las dimensiones del vector de probabilidades.
    print("Dimensiones de la probabilidad de los tokens:", probs.shape)

    # Ordenamos los tokens según su probabilidad de mayor a menor.
    sorted_probs = torch.argsort(probs, dim=-1, descending=True)

    # Mostramos los 'best' tokens más probables y su probabilidad en porcentaje.
    print({tokenizer_gptneo.decode(token): f"{prob.cpu().numpy() * 100:.2f}%"
           for token, prob in zip(sorted_probs[:best], probs[sorted_probs[:best]])})

Dimensiones de la entrada: torch.Size([1, 3])
Dimensiones de la salida: torch.Size([1, 3, 50257])
Dimensiones del último token de la secuencia: torch.Size([50257])
Dimensiones de la probabilidad de los tokens: torch.Size([50257])
{',': '57.87%', ' there': '5.13%', ' I': '4.05%', ' a': '4.02%', '\n': '2.72%', ' in': '2.57%', ' (': '2.14%', '…': '1.71%', '...': '1.52%', ' the': '1.25%'}


### **Cometario sobre Generación de Texto con los Modelos GPT-2 y GPT-Neo**

Podemos ver que ambos modelos dan opciones diferentes y ambos sugieren como primera opción `','`, pero podemos notar que:  

- Con **GPT-2**, la probabilidad es del **46.95%**.  
- Con **GPT-Neo**, la probabilidad es del **57.87%**.  

Esto significa que nuestro modelo **GPT-Neo** le da mayor *confianza* a sus predicciones; por lo que será mucho mejor en generación de texto que **GPT-2**. Es realmente lo essperado.  Por lo que, de aqui en adelante trabajaremo con el Modelo *GPT-Neo*; la razón ya la hicimos evidente.

## **8. Implementando una función de generación de texto en Inglés con GPT-Neo**

Ahora, la idea es que este modelo nos sirva para generar texto de forma recurrente e incremental. En la última capa de los modelos tipo GPT encontrarémos un tensor con forma $(b, s, v)$, donde:

- $b$: Es el tamaño del bache, o la cantidad de secuencias a procesar.
- $s$: Es la longitud de la secuencia de entrada.
- $v$: Es el tamaño del vocabulario del modelo, cuantos tokens soporta.

Pero este es el tensor de salida, por qué tiene la forma de la secuencia de entrada?, porque cada posición en la salida corresponde a la la predicción del siguiente token de esa posición en la secuencia de entrada. En otras palabras, lo que obtenemos como predicción, es una secuencia de igual tamaño a la de entrada, movida un token hacia adelante, lo que efectivamente nos predice un solo token a la vez y por ende, el token que nos insteresa, es el último.

Lo que obtenemos en este tensor es además los logits de TODO el vocabulario del modelo, con los cuales podemos calcular las probabilidades de que cada uno sea el que continue en la secuencia. Hay varias formas de decodificar el siguiente token, la más fácil de implementar sería una decodificación codiciosa (greedy) del siguiente token, que consiste simplemente en seleccionar el token con la probabilidad más alta. Este es un enfoque simple y efectivo para algunos casos, pero al mismo tiempo sufre de poca variabilidad e incluso puede caer en generación repetitiva.

Otra opción es el muestreo, ya que justamente podemos obtener probabilidades del siguiente token, lo más lógico sería muestrear con esas opciones de probabilidad, de este modo podemos obtener mayor diversidad a la hora de generar el texto, al costo eso si de que haya mayor aleatoridad ya que se le daría la oportunidad a incluso tokens con baja probabilidad, de ser seleccionados.

Otra opción podría ser un balanceo entre una decodificación greedy y una por muestreo, en función de otro hiperparámetro que podemos definir. Esta sería una técnica muy común en el contexto de Reinforcement Learning llamade e-greedy. Se hace la aclaración de que en este ejemplo no harémos nada de RL, solamente se hace mención de esta técnica para balancear entre explotación y exploración.

In [ ]:
def generate(
        model: nn.Module,  # Modelo de lenguaje causal (GPT-2, GPT-Neo, etc.)
        tokenizer: PreTrainedTokenizerBase,  # Tokenizador correspondiente al modelo
        start: str,  # Texto inicial para generar la secuencia
        max_length: int = 100,  # Longitud máxima de tokens a generar
        eps: float = 0.5,  # Parámetro e-greedy para balance entre greedy y muestreo
        top_n: int = 5,  # Número de mejores opciones a considerar para análisis
        return_iterations: bool = False,  # Si True, devuelve detalle de cada iteración
        device: str = "cpu") -> Tuple[str, Optional[pd.DataFrame]]:  # Dispositivo (CPU/GPU) y tipo de retorno

    output = [start]  # Inicializamos la salida con el texto inicial
    iterations = []  # Lista para guardar detalles de cada iteración (opcional)

    with torch.no_grad():  # Desactivamos gradientes, solo generamos texto
        input_ids = tokenizer_gptneo(output[-1], return_tensors='pt')['input_ids'].to(device)  # Tokenizamos y pasamos a GPU/CPU

        for _ in range(max_length):  # Iteramos hasta generar max_length tokens
            logits = model_gptneo(input_ids=input_ids).logits  # Predicción cruda (logits) del siguiente token
            probs = torch.softmax(logits[0, -1, :], dim=-1)  # Convertimos logits en probabilidades
            sorted_tokens = torch.argsort(probs, dim=-1, descending=True)  # Ordenamos tokens por probabilidad descendente

            # Política e-greedy: balance entre greedy (tomar el mejor token) y muestreo
            if np.random.random_sample(1)[0] < eps:
                next_token = sorted_tokens[0].unsqueeze(dim=0)  # Tomamos el token más probable (greedy)
            else:
                next_token = torch.multinomial(probs, 1)  # Muestreamos token según su distribución de probabilidad

            if return_iterations:  # Guardamos detalle de la iteración para análisis
                iteration = {'input': ''.join(output)}  # Texto generado hasta ahora
                best_n = sorted_tokens[:top_n].cpu().tolist()  # Top N tokens más probables
                # Creamos diccionario con token decodificado y su probabilidad
                choices = {f'Choice #{choice+1}': f'{tokenizer.decode(token)} ({prob:.4f})'
                           for choice, (token, prob) in enumerate(zip(best_n, probs[best_n].cpu().tolist()))}
                iteration.update(choices)  # Actualizamos iteración con las opciones
                iterations.append(iteration)  # Guardamos la iteración en la lista

            output.append(tokenizer.decode(next_token))  # Decodificamos el token y lo agregamos al output
            input_ids = torch.cat([input_ids, next_token.unsqueeze(dim=0)], dim=-1)  # Actualizamos input_ids para siguiente predicción

        output_text = ''.join(output)  # Convertimos lista de tokens en un string final

        if not return_iterations:
            return output_text, None  # Retornamos solo el texto
        else:
            df = pd.DataFrame(iterations)  # Creamos DataFrame con detalle de cada iteración
            return output_text, df  # Retornamos texto + DataFrame

### **8.1. Generación de Texto usando la Función definida en el Paso 8**

Vamos a usar la funcion *Generate* definida en el Paso 8, para generar texto usando el Modelo *GPT-Neo* y su Tokenizador (*tokenizer_gpteno*). Veamos como le va haciendo predicciones en la secuencia del texto.

#### **8.1.1. Uso de la Funcion con eps=1.0**

Veamos como le va generando texto usando un eps=1.0.

In [ ]:
# Generamos texto usando la función 'generate' con nuestro modelo GPT-Neo
# max_length=15 limita la generación a 15 tokens
# eps=1.0 significa que siempre tomaremos el token más probable (modo greedy)
# top_n=10 nos permite analizar las 10 opciones más probables por iteración
# return_iterations=True guarda información de cada paso en un DataFrame
output_text, iterations_df = generate(
    model_gptneo,        # Modelo GPT-Neo cargado previamente
    tokenizer_gptneo,    # Tokenizador correspondiente a GPT-Neo
    text,                # Texto inicial para comenzar la generación
    max_length=15,       # Número máximo de tokens a generar
    eps=1.0,             # Política greedy (siempre el token más probable)
    top_n=10,            # Número de opciones a mostrar por iteración
    return_iterations=True,  # Retornar detalle de cada iteración
    device=device        # Dispositivo (CPU o GPU) donde se realiza la generación
)

# Mostramos el texto generado completo
print("Este es el texto completo generado en Inglés por el Modelo GPT-Neo: \n", output_text)

# Mostramos las primeras 15 filas del DataFrame de iteraciones para analizar decisiones del modelo
print("\n Muestra de Texto generado e Iteraciones realizadas por el Modelo GPT-Neo")
iterations_df.head(15)


Este es el texto completo generado en Inglés por el Modelo GPT-Neo: 
 Once upon time, a young man named John Smith was walking down the street in London.

 Muestra de Texto generado e Iteraciones realizadas por el Modelo GPT-Neo


,input,Choice #1,Choice #2,Choice #3,Choice #4,Choice #5,Choice #6,Choice #7,Choice #8,Choice #9,Choice #10
0,Once upon time,", (0.5787)",there (0.0513),I (0.0405),a (0.0402),\n (0.0272),in (0.0257),( (0.0214),… (0.0171),... (0.0152),the (0.0125)
1,"Once upon time,",a (0.1779),there (0.1392),I (0.0824),the (0.0691),in (0.0604),when (0.0449),an (0.0208),we (0.0142),one (0.0116),my (0.0098)
2,"Once upon time, a",young (0.0531),man (0.0530),group (0.0312),certain (0.0248),little (0.0189),girl (0.0174),few (0.0169),small (0.0155),woman (0.0151),guy (0.0150)
3,"Once upon time, a young",man (0.2066),girl (0.1145),boy (0.0890),woman (0.0853),", (0.0234)",couple (0.0182),lady (0.0159),and (0.0085),writer (0.0064),guy (0.0062)
4,"Once upon time, a young man",named (0.2086),was (0.0599),from (0.0473),came (0.0376),", (0.0372)",in (0.0350),and (0.0301),with (0.0283),by (0.0258),who (0.0234)
5,"Once upon time, a young man named",John (0.0158),David (0.0157),Daniel (0.0107),Michael (0.0105),Jack (0.0104),Adam (0.0088),James (0.0088),Henry (0.0082),Peter (0.0081),William (0.0077)
6,"Once upon time, a young man named John",Smith (0.0385),was (0.0281),had (0.0118),Henry (0.0116),Doe (0.0115),came (0.0114),Lennon (0.0106),Wayne (0.0100),D (0.0087),", (0.0083)"
7,"Once upon time, a young man named John Smith",was (0.1425),", (0.0596)",went (0.0335),came (0.0321),had (0.0318),walked (0.0263),( (0.0247),and (0.0236),decided (0.0215),set (0.0214)
8,"Once upon time, a young man named John Smith was",walking (0.0623),a (0.0387),sent (0.0339),in (0.0331),born (0.0270),traveling (0.0220),living (0.0201),on (0.0200),looking (0.0156),working (0.0151)
9,"Once upon time, a young man named John Smith w...",down (0.1927),along (0.1783),through (0.1599),home (0.1411),in (0.0471),on (0.0370),to (0.0318),with (0.0243),alone (0.0209),the (0.0203)


#### **Comentario sobre el punto 8.1.1. Uso eps=1.0**
Observamos como el input progresa a la vez que las opciones de tokens
que hay. Sin importar cuantas veces invoquemos a la función con los mismos parámetros, siempre vamos a obtener los mismos resultados.

#### **8.1.2. Uso de la Función con un eps=0.5**

Veamos ahora como le va a la función generando texto usando un eps=0.5

#### **8.1.2.1. Primera Prueba de la Funcion con eps=0.5**

In [ ]:
### PRUEBA NRO. 1

# Generamos texto usando la función 'generate' previamente definida.
# Le pasamos el modelo GPT-Neo y su tokenizador, además del texto inicial 'text'
# max_length=100 limita la longitud máxima de la secuencia generada a 100 tokens
# eps=0.5 indica que estamos usando una estrategia e-greedy balanceada para seleccionar tokens
# device=device asegura que el modelo se ejecute en la GPU si está disponible, o CPU en caso contrario
output_text, _ = generate(
    model_gptneo,        # modelo de lenguaje GPT-Neo
    tokenizer_gptneo,    # tokenizador asociado al modelo GPT-Neo
    text,                # texto inicial a partir del cual generamos la secuencia
    max_length=100,      # longitud máxima de la secuencia generada
    eps=0.5,             # probabilidad de elegir el token más probable vs. muestreo aleatorio
    device=device        # dispositivo de cómputo (CPU o GPU)
)

# Mostramos el texto generado en pantalla
print(output_text)

Once upon time, a young woman walked into a bar and asked for a drink. As she was about to place her order, the bartender asked ��What did you order?��

��What drink?�� she replied.

��Herbal tea,�� replied the bartender.

��Herbal tea,�� she repeated.

��You��re not from a land called England.��

��I was told that�


#### **8.1.2.2. Segunda Prueba de la Funcion con eps=0.5**

In [ ]:
### PRUEBA NRO. 2

# Generamos texto usando la función 'generate' previamente definida.
# Le pasamos el modelo GPT-Neo y su tokenizador, además del texto inicial 'text'
# max_length=100 limita la longitud máxima de la secuencia generada a 100 tokens
# eps=0.5 indica que estamos usando una estrategia e-greedy balanceada para seleccionar tokens
# device=device asegura que el modelo se ejecute en la GPU si está disponible, o CPU en caso contrario
output_text, _ = generate(
    model_gptneo,        # modelo de lenguaje GPT-Neo
    tokenizer_gptneo,    # tokenizador asociado al modelo GPT-Neo
    text,                # texto inicial a partir del cual generamos la secuencia
    max_length=100,      # longitud máxima de la secuencia generada
    eps=0.5,             # probabilidad de elegir el token más probable vs. muestreo aleatorio
    device=device        # dispositivo de cómputo (CPU o GPU)
)

# Mostramos el texto generado en pantalla
print(output_text)

Once upon time they went to Africa to give a Robert Frost lecture.

The talk was in the

The lecture was in New York City.

The Robert Frost lecture is pretty much like any lecture you might hear today. It��s a bit long, but it��s not a linear lecture with a lot of information and then a bit of a back-and-forth with an audience.

Instead, it��s a succession of short talks, each of which is


#### **8.1.2.3. Tercera Prueba de la Funcion con eps=0.5**

In [ ]:
## PRUEBA NRO. 3

# Generamos texto usando la función 'generate' previamente definida.
# Le pasamos el modelo GPT-Neo y su tokenizador, además del texto inicial 'text'
# max_length=100 limita la longitud máxima de la secuencia generada a 100 tokens
# eps=0.5 indica que estamos usando una estrategia e-greedy balanceada para seleccionar tokens
# device=device asegura que el modelo se ejecute en la GPU si está disponible, o CPU en caso contrario
output_text, _ = generate(
    model_gptneo,        # modelo de lenguaje GPT-Neo
    tokenizer_gptneo,    # tokenizador asociado al modelo GPT-Neo
    text,                # texto inicial a partir del cual generamos la secuencia
    max_length=100,      # longitud máxima de la secuencia generada
    eps=0.5,             # probabilidad de elegir el token más probable vs. muestreo aleatorio
    device=device        # dispositivo de cómputo (CPU o GPU)
)

# Mostramos el texto generado en pantalla
print(output_text)

Once upon time, a travel agent was in the middle of something

Nothing gives me more pleasure than to travel. After all, traveling isn��t only about seeing new places, it is often about experiencing new cultures, meeting new people, and having unforgettable experiences.

But, that��s not all. Travel is about meeting people, and the best way to do that is through the people you meet. And, we at Travel Agent, understand that meeting people is an extra-special experience


### **Conclusiones sobre el efecto de `eps` en la generación de texto**

- **Con `eps=1.0`**: El modelo siempre elige el token con la mayor probabilidad (*greedy*). Esto hace que el texto generado sea más predecible y coherente, pero puede resultar un poco repetitivo y menos creativo, ya que no explora otras opciones de tokens.

- **Con `eps=0.5`**: El modelo mezcla la selección *greedy* con muestreo aleatorio, eligiendo a veces tokens menos probables según su distribución de probabilidades. Esto introduce mayor diversidad y creatividad en el texto, aunque como vimos genera en las pruebas incoherencias en algunas partes del texto.


## **9. Generando texto con las utilidades del modelo**

Ahora, la clase de Huggingface implementa la función `generate` que hace la labor de generación por nosotros, incluyendo las opciones de muestreo y explotación como hemos observado. Solo que además permite otra serie de parámetros y opciones para controlar la generación de texto. Para más información se recomienda estudiar:

- [Natural Language Processing with Transformers: Building Language Applications With Hugging Face](https://www.amazon.com/Natural-Language-Processing-Transformers-Applications/dp/1098103246), Capitulo 5
- https://huggingface.co/docs/transformers/v4.41.3/en/main_classes/text_generation#transformers.GenerationConfig
- https://huggingface.co/docs/transformers/v4.41.3/en/main_classes/text_generation#transformers.GenerationMixin.generate

In [ ]:
# Generamos texto a partir del modelo GPT-Neo usando los tokens de entrada
# 'pad_token_id' se define como el token EOS para que el modelo sepa cuándo rellenar
# 'max_length' indica la longitud máxima del texto a generar (100 tokens)
# 'do_sample=True' activa el muestreo aleatorio para aumentar la diversidad de salida
# 'temperature=0.5' ajusta la aleatoriedad del muestreo: valores más bajos hacen el texto más predecible
# 'top_k=0' indica que no limitamos la selección a los k tokens más probables, se usan todos
output = model_gptneo.generate(tokens, pad_token_id=tokenizer_gptneo.eos_token_id, max_length=100, do_sample=True, temperature=0.5, top_k=0)

# Decodificamos los IDs de tokens generados a texto legible y lo imprimimos
print(tokenizer_gptneo.decode(output[0]))

The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Once upon time, there was a time when a man named Adolf Hitler wanted to be president. He was a follower of the philosophy of National Socialism, which was a blend of socialism and fascism. The Nazis were determined to bring about the overthrow of the modern world order and replace it with a new form of government that would be more in line with their ideals.

Hitler was a man who believed in a strong central government. He was also a man who was well-liked in his


## **10. Fine tuning con el modelo GPT-Neo**

Ahora, intentemos hacer fine tuning a nuestro modelo. Intentemos entrenarlo en un corpus de chistes en idioma Inglés y ver como la narrativa de su salida cambia.

##### Nota
El objetivo es observar como cambia la generación del texto una vez lo entrenamos en un conjunto especializado, que es en este caso chistes (jokes) en Ingles.

##### Referencia
Usamos un dataset disponible en Kaggle: https://www.kaggle.com/datasets/abhinavmoudgil95/short-jokes. Este dataset esta compuesto por 231,657 chistes (jokes) en idioma Inglés. La longitud de los chistes va desde los 10 a los 200 caracteres, lo que es común dado que son chistes, y su narrativa no suele ser extensa.

In [ ]:
# 1. Configuramos las credenciales de Kaggle para poder descargar datasets desde la API
# Aquí se asigna el nombre de usuario y la clave de Kaggle a variables de entorno
os.environ["KAGGLE_USERNAME"] = "octavioguerram"
os.environ["KAGGLE_KEY"] = "f05dbf9bc70ef06f0cea6965c7b01d14"

# 2. Usamos la interfaz de línea de comandos de Kaggle para descargar el dataset
# '-d' indica el identificador del dataset en Kaggle
# '-p ./data' define la ruta local donde se guardará el dataset
# '--unzip' descomprime automáticamente el archivo descargado
!kaggle datasets download -d abhinavmoudgil95/short-jokes -p ./data --unzip

# 3. Cargamos directamente el dataset usando la librería `datasets` de Hugging Face
# Se indica que el archivo es CSV y su ubicación local
dataset = load_dataset("csv", data_files="./data/shortjokes.csv")

# Mostramos información general del dataset cargado
print(dataset)

# Imprimimos la primera fila del conjunto de entrenamiento para ver su contenido
print(dataset['train'][0])


Dataset URL: https://www.kaggle.com/datasets/abhinavmoudgil95/short-jokes
License(s): DbCL-1.0
  0% 0.00/9.82M [00:00<?, ?B/s]
100% 9.82M/9.82M [00:00<00:00, 1.08GB/s]


Generating train split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['ID', 'Joke'],
        num_rows: 231657
    })
})
{'ID': 1, 'Joke': '[me narrating a documentary about narrators] "I can\'t hear what they\'re saying cuz I\'m talking"'}


## **11. Reducción del Dataset para Entrenamiento del Modelo GPT-Neo**

Vamos a reducir el Dataset de Entrenamiento que escogimos que tiene 231,657 registros a 20 mil. Esto lo hacemos para reducir el tiempo de entrenamiento del Modelo GPT-Neo.

In [ ]:
# Reduzcamos la cantidad de registros del dataset
# Esto se hace para disminuir el tiempo de entrenamiento y poder hacer pruebas rápidas
num_reg = 20000  # Definimos el número de registros que queremos mantener

# Mezclamos aleatoriamente (shuffle) el dataset de entrenamiento para que la selección sea representativa
# La semilla 'seed=42' asegura reproducibilidad, es decir, que siempre se obtenga el mismo orden al ejecutar el código
# Luego, seleccionamos solo los primeros 'num_reg' registros para reducir el tamaño del dataset
dataset['train'] = dataset['train'].shuffle(seed=42).select(range(num_reg))

### **11.1. Verificación del Nuevo Dataset reducido y su tamaño**

Confirmemos con el código en la celda abajo, que ahora tenemos un Dataset reducido a 20 mil Registros.

In [ ]:
# Mostramos el contenido actual del dataset
# Esto nos permite ver cuántos registros tenemos y cómo está estructurado después de la reducción
dataset

DatasetDict({
    train: Dataset({
        features: ['ID', 'Joke'],
        num_rows: 20000
    })
})

### **11.2. Muestra de un registro del Nuevo Dataset Reducido**

Veamos el primer registro del Dataset reducido a 20 mil Registros. Lo hacemos para confirmar que sigue manteniendo el contenido esperado de chistes en Idioma Inglés.

In [ ]:
# Muestra en pantalla del primer registro en el Dataset reducido a 20 Mil Registros.
dataset['train'][0]

{'ID': 218585, 'Joke': 'What is long, hard, and has cum in it? Cucumber'}

### **11.3. Muestra del Nuevo Dataset reducido en formato Panda Dataframe**

Veamos el Dataset reducido a 20 Mil Registros pero usando Panda Dataframes. Es al final una manera mas "estética" de verlo.

In [ ]:
# Cambiamos el formato del dataset a Pandas para poder manipularlo más fácilmente
dataset.set_format('pandas')

# Convertimos la partición 'train' del dataset a un DataFrame de Pandas
df = dataset['train'].to_pandas()

# Mostramos las primeras 10 filas del DataFrame para inspeccionar los registros
# queremos verlos en una mejor forma
df.head(10)


,ID,Joke
0,218585,"What is long, hard, and has cum in it? Cucumber"
1,196035,"(Grandpa joke) If seagulls fly over the sea, w..."
2,5293,I always heard it was OK to date a nun... as l...
3,14148,I accidentally drank a bottle of ink. The doct...
4,19827,If anyone finds 786 barely used tubes of Chaps...
5,64124,There's a bizarre irony to jokes about Ebola g...
6,101633,The only laid I'm getting... is laid off.
7,107397,"The Chinese police made an arrest today, but h..."
8,146886,Why did Adele cross the road? To say hello fro...
9,176560,friend: have u accepted jesus christ as yr sav...


### **11.4. Cantidad de Palabras Máxima, Mínima y Media para el Dataset Reducido**

Veamos ahora usando el codigo en la celda siguiente:

- Cual es la máxima cantidad de palabras que tiene el Dataset Reducido.
- Cual es la mínima cantidad de palabras que tiene el Dataset Reducido.
- Cual es la media de palabras que tiene el Dataset Reducido.

*Nota:* Recordemos que el Dataset que estamos usando fué reducido a 20 Mil Registros y contiene chistes (jokes) en el Idioma Inglés

In [ ]:
# Creamos una nueva columna que almacena la cantidad de palabras de cada chiste
df['Palabras por chiste'] = df['Joke'].str.split().apply(len)

# Mostramos la cantidad más larga de palabras que hay en el dataset
print("Chiste con mayor cantidad de palabras: ", df['Palabras por chiste'].max())

# Mostramos la cantidad más pequena de palabras que tiene el dataset
print("Chiste con menor cantidad de palabras: ", df['Palabras por chiste'].min())

# Calculamos la mediana de palabras por chiste, que nos da una idea del promedio típico de palabras
print("Palabras Media por Chiste: ", df['Palabras por chiste'].median())


Chiste con mayor cantidad de palabras:  46
Chiste con menor cantidad de palabras:  1
Palabras Media por Chiste:  17.0


### **Comentarios sobre La Longitud de Palabras del Dataset**

En este análisis podemos observar que la mediana de longitud de los chistes, en términos de palabras, es de 17. Este valor resulta lógico, ya que por su propia naturaleza los chistes suelen ser textos cortos y directos.

Sin embargo, también debemos tener en cuenta que el corpus original utilizado para entrenar el modelo GPT-Neo estaba compuesto por textos de características muy diferentes a los chistes (por ejemplo, libros, artículos o páginas web). Por esta razón, la calidad de los resultados al usar directamente el modelo preentrenado podría no ser óptima, a menos que se realice un ajuste fino adaptado a nuestro dataset.

El análisis de la longitud de palabras por chiste nos resulta muy útil porque nos permite establecer un valor apropiado para el parámetro max_length durante el entrenamiento. Esto es importante, ya que:

- Si usamos un valor demasiado grande (por ejemplo, 512 o 256), la mayoría de las secuencias tendrán que rellenarse con muchos tokens de padding, lo que haría que el entrenamiento fuera más lento e ineficiente.

- En cambio, al conocer las longitudes máxima, mínima y mediana, podemos seleccionar un max_length adecuado que represente bien a los datos, evitando exceso de padding y mejorando la eficiencia del modelo.

## **12. Preprocesamiento del dataset con tokenización**

En este parte del código definimos una función de preprocesamiento que se encarga de tokenizar los chistes en Inglés utilizando el modelo GPT-Neo. La función aplica truncamiento y padding para asegurar que todas las secuencias tengan la misma longitud máxima (max_len), lo cual es necesario para que el entrenamiento sea más eficiente y consistente.

In [ ]:
# En este bloque definimos una función de preprocesamiento para el dataset.
# La idea es tokenizar cada chiste utilizando el tokenizador de GPT-Neo.
#
# 1. La función externa `preprocess_function(max_len)` recibe como parámetro
#    la longitud máxima (`max_len`) que queremos para cada secuencia.
#
# 2. Dentro de ella, se define la función interna `_preprocess_function(examples)`,
#    la cual toma como entrada un conjunto de ejemplos del dataset.
#
# 3. Esta función aplica el tokenizador sobre la columna "Joke", truncando los
#    textos que sean demasiado largos y rellenando con padding hasta la longitud
#    máxima definida.
#
# Con esto garantizamos que todas las secuencias tengan la misma longitud,
# lo cual es un requisito indispensable para poder entrenar el modelo de forma
# eficiente en tareas de generación de texto.
def preprocess_function(max_len):
    def _preprocess_function(examples):
        return tokenizer_gptneo(
            examples['Joke'],
            max_length=max_len,
            truncation=True,
            padding='max_length'
        )
    return _preprocess_function

## **13. Restablecimiento del Formato Original del Dataset**

Con este comando restablecemos el formato original del dataset. Antes lo habíamos convertido a 'pandas' para manipularlo como un DataFrame. Al usar reset_format(), el dataset vuelve a estar en el formato por defecto de Hugging Face.Lo hacemos para poderlo usar en el entrenamiento del Modelo GPT-Neo que tenemos.

In [ ]:
dataset.reset_format()

## **14. Tokenización del Dataset Reducido**

Tomamos el Dataset Reducido que contiene 20 mil Registros de chistes en inglés, que usaremos para entrenar un modelo GPT-Neo. Pero para hacerlo hay que Tokenizarlo. El codigo en la celda permite definir la longitud máxima de las secuencias y es clave este valor, que debe escoger acorde a los valores de longitud de palabra maxima, minima y media del conjunto de datos. Enseguida, aplicamos la tokenización de los chistes, y proseguimos con una división del conjunto de datos en dos subconjuntos: Entrenamiento y Prueba. Estos dos subconjuntos los adaptamos al formato de PyTorch. Y como final, configuramos un Data Collator adecuado para modelos causales, asegurando que el entrenamiento se realice de manera correcta.


In [ ]:
# Definimos la longitud máxima de las secuencias que procesará el modelo.
max_long = 64

# Aplicamos la función de preprocesamiento al conjunto de datos de entrenamiento.
# Esto tokeniza los chistes con la longitud máxima definida y los prepara para el modelo.
tokenized_dataset = dataset["train"].map(preprocess_function(max_long), batched=True)

# Dividimos el dataset en entrenamiento y prueba.
# Con test_size=0.1 reservamos el 10% de los datos para evaluar el rendimiento del modelo.
tokenized_dataset = tokenized_dataset.train_test_split(test_size=0.1)

# Ajustamos el formato de los datos para que solo contenga lo que el modelo necesita:
# los IDs de los tokens y la máscara de atención, en formato tensor de PyTorch.
tokenized_dataset.set_format(type="torch", columns=["input_ids", "attention_mask"])

# Creamos un Data Collator, que se encarga de organizar los lotes de datos durante el entrenamiento.
# En este caso es importante indicar mlm=False porque GPT-Neo es un modelo causal
# y no funciona con enmascarado de palabras como los modelos tipo BERT.
data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer_gptneo,
    mlm=False   # GPT-Neo es causal LM, no Masked LM
)

## **15. Configuración del Entrenador con Hugging Face Trainer**

En esta sección definimos los parámetros globales de entrenamiento (épocas, tasa de aprendizaje, tamaño de lote, etc.) y configuramos el objeto `Trainer`.  
Este componente nos permitirá entrenar y evaluar el modelo GPT-Neo de manera estructurada, aprovechando las utilidades que ofrece la librería de Hugging Face para la gestión de datasets, métricas y checkpoints.


In [ ]:
# Definimos el tamaño de los lotes de entrenamiento y validación
batch_size = 8

# Calculamos cada cuántos pasos se mostrará el log, en función del tamaño del dataset
logging_steps = len(tokenized_dataset['train']) // batch_size

# Establecemos los parámetros globales de entrenamiento para Hugging Face Trainer
training_args = TrainingArguments(
    output_dir='./hf-gpt',              # Carpeta donde se guardarán los resultados y checkpoints
    overwrite_output_dir=True,          # Permite sobrescribir el directorio de salida si ya existe
    num_train_epochs=3,                 # Número de épocas de entrenamiento
    learning_rate=2e-5,                 # Tasa de aprendizaje inicial
    per_device_eval_batch_size=batch_size,   # Tamaño de lote para evaluación
    per_device_train_batch_size=batch_size,  # Tamaño de lote para entrenamiento
    weight_decay=0.01,                  # Factor de regularización para evitar sobreajuste
    evaluation_strategy='epoch',        # Evaluar el modelo al final de cada época
    save_strategy='epoch',              # Guardar checkpoints al final de cada época
    load_best_model_at_end=True,        # Cargar el mejor modelo tras el entrenamiento
    disable_tqdm=False,                 # Mantener la barra de progreso activa
    logging_steps=logging_steps,        # Frecuencia de registro de métricas
    logging_strategy='epoch',           # Registrar métricas cada 'logging_steps' pasos
    report_to='none'                   # No enviar métricas a plataformas externas
)

# Definimos el entrenador de Hugging Face,
# indicando el modelo, los parámetros, los datasets, el tokenizador y el collator
trainer = Trainer(
    model=model_gptneo,                     # El modelo GPT-Neo preentrenado que ajustaremos
    args=training_args,                     # Los parámetros globales definidos arriba
    train_dataset=tokenized_dataset['train'],   # Subconjunto de datos para entrenamiento
    eval_dataset=tokenized_dataset['test'],     # Subconjunto de datos para evaluación
    tokenizer=tokenizer_gptneo,             # Tokenizador correspondiente al modelo
    data_collator=data_collator             # Objeto que organiza los datos en lotes
)

## **16. Ejecución y Medición del Entrenamiento del Modelo**

Con el código en la celda abajo damos inicio al proceso de entrenamiento de GPT-Neo utilizando el objeto `Trainer` previamente configurado.  
Además, usamos  `%%time` de Jupyter/Colab para medir automáticamente el tiempo total de ejecución, lo que nos permite evaluar la eficiencia del entrenamiento bajo los parámetros definidos.


In [ ]:
# Midamos y mostremos cuánto tiempo tarda en ejecutarse el entrenamiento basado en el
# tiempo de ejecución (running) de esta celda de código.
%%time

# Iniciamos el proceso de entrenamiento del modelo con los parámetros, datasets y métricas que configuramos antes.
trainer.train()

Epoch,Training Loss,Validation Loss
1,2.488100,2.536429
2,1.382400,2.674432
3,0.719300,2.895709


There were missing keys in the checkpoint model loaded: ['lm_head.weight'].


CPU times: user 21min 6s, sys: 45.3 s, total: 21min 51s
Wall time: 23min 13s


TrainOutput(global_step=6750, training_loss=1.5299256908275463, metrics={'train_runtime': 1393.3458, 'train_samples_per_second': 38.756, 'train_steps_per_second': 4.844, 'total_flos': 2.5058526363648e+16, 'train_loss': 1.5299256908275463, 'epoch': 3.0})

## **17. Pruebas con el Modelo GPT-Neo Entrenado**

Veamos como le va al modelo GPT-Neo entrenado usando "Fine Tuning" y con el Dataset que seleccionamos de chistes en el idioma Inglés.

### **17.1. Prueba Nro. 1**

In [ ]:
model_trained = trainer.model

output = model_trained.generate(           # Le pedimos al modelo GPT-Neo entrenado que genere texto a partir de los tokens de entrada
    tokens,                                # Tokens de entrada que representan el texto inicial
    pad_token_id=tokenizer_gptneo.eos_token_id,  # Definimos el token de padding como el token de fin de secuencia
    max_length=100,                        # Limitamos la generación a 100 tokens como máximo
    do_sample=True,                        # Activamos muestreo para que la salida sea creativa y no siempre igual
    temperature=0.8                        # Controlamos la “creatividad” del modelo; 0.8 da un balance entre coherencia y variedad
)

# Decodificamos los tokens generados a texto legible y lo mostramos en pantalla
print(tokenizer_gptneo.decode(output[0]))


Once upon time, you take the bus to the city and then you go to a big haunted house. It's not the end of the world. I'll just be there in a couple of hours. Just like usual *dinner party* *dinner party* Okay, that was cheesy. I'm out of here. -*hangs head* *faints at foot* -*just kidding* *flips on bus and screams* -*I'm still here* -*


### **17.2. Prueba Nro. 2**

In [ ]:
output_text, _ = generate(model_trained, tokenizer_gptneo, text, max_length=100, eps=0.2, device=device)
print(output_text)

Once upon time I asked a married couple what their favorite fairy tale was, happily married they said "It's such a pain in the ass to get." I said "I'm not married, your favorite were to get?" They replied "fully of each". I'm own. :D. I'm sorry I didn't get your joke :D. I like your style. B relay! I'll see myself out. EDIT: Never mind, I was fired. :'(.OMG Let's hear


Nop parecen ser chistes y si lo son realmente no son muy graciosos. El ttexto generado en la Prueba Nro. 1, da un poco de gracia, al considerar que es un sueño momentaneo de alguien que va en un bus. El generado en la Prueba Nro. 2, honestamente no le encontramos mucho sentido, pero bueno lo intento el Modelo entrenado.

Lo interesante al final, es que con el entrenamiento, logramos darle un giro al estilo de generación de texto del Modelo GPT-Neo, con el que fue entrenado originalmente. Es decir, que definitivamente el conjunto de datos logra influir al Modelo Original. Y será algo que tendremos que tener en cuenta a la hora de *utilizar* y hacer *fine tuning* a Modelos de Lenguaje, como en este caso el que seleccionamos **GPT-Neo**



## **Conclusiones**

- Al trabajar con GPT-Neo en este Google Colab Notebook, pudimos evidenciar  el dilema entre exploración y explotación en la generación de texto, que tambien fue mostrado en el Google Colab Notebook original con GPT-2. Cuando usamos *greedy search*, el modelo resultó más preciso y coherente, pero también repetitivo y monótono. Mientras que, con técnicas de *sampling*, el texto se vuelve más creativo y diverso, aunque con mayor riesgo de incoherencias o alucinaciones, lo que lo hace divertido de alguna manera. Es como alguien aprendiendo a escribir y con un vocabulario muy básico. Al final, esto nos dice que que es importante ajustar el balance según la tarea específica que tenemos propuesta.

- Después de los prácticas hechas con BERT y GPT. Como equipo que ha desarrollado ambas, entendemos que son claramente modelos transformers, pero con objetivos distintos: BERT se entrena de manera bidireccional, aprendiendo a “rellenar huecos” en frases, lo que lo hace muy bueno para tareas de comprensión como clasificación o respuesta a preguntas. En cambio, GPT es un modelo autoregresivo, que predice la siguiente palabra en una secuencia, lo cual lo hace más fuerte en generación de texto. Lo resumimos en que: BERT entiende mejor, mientras que GPT escribe mejor.

- Nos parecio interesante, ver la diferencia al generar texto del Modelo GPT-Neo, que usamos en este ejercicio como equipo  vs GPT-2, que venia en el Notebook original. Revisando la documentación, GPT-2, fue entrenado con 1.5 Billones de Parámetros; mientras que, GPT-Neo, fue entrenado con 2.7 Billones de Parámetros. Y eso fué notorio en los puntos 7.1 y 7.2 desarrollados en este Google Colab Notebook. La probabilidad entregada por el modelo GPT-Neo vs la que entregaba GPT-2, marca la diferencia a la hora de generación de texto y fue lo que nos llevo a elegirlo para realizar el resto de la práctica.

- Nuevamente se hace visible la combinación de la máxima longitud de la secuencia, tamaño del batch, cantidad de epochs, dataset seleccionado y si se encuentra balanceado y curado. Todo esto, contribuyen a que el Modelo tenga datos adecuados y haga un adecuado entrenamiento, que en nuestro caso produzca mejores chistes en ingles. Es decir que se vuelve: entrenar, probar, ajustar, entrenar, hasta que la producción sea lo esperado.

